src/Compilers/Server/VBCSCompiler/NamedPipeClientConnectionHost.cs

```diff
@@ -13,50 +13,208 @@
-            // starting listening.
-            CompilerServerLogger.Log("Constructing pipe '{0}'.", _pipeName);
-            var pipeStream = NamedPipeUtil.CreateServer(_pipeName);
-            CompilerServerLogger.Log("Successfully constructed pipe '{0}'.", _pipeName);
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                NamedPipeServerStream? pipeStream = null;
 
-            CompilerServerLogger.Log("Waiting for new connection");
-            await pipeStream.WaitForConnectionAsync(cancellationToken).ConfigureAwait(false);
-            CompilerServerLogger.Log("Pipe connection detected.");
+                try
+                {
+                    // Create the pipe and begin waiting for a connection. This 
+                    // doesn't block, but could fail in certain circumstances, such
+                    // as Windows refusing to create the pipe for some reason 
+                    // (out of handles?), or the pipe was disconnected before we 
+                    // starting listening
+                    CompilerServerLogger.Log($"Constructing pipe and waiting for connections '{pipeName}'");
+                    pipeStream = NamedPipeUtil.CreateServer(pipeName);
 
-            if (Environment.Is64BitProcess || MemoryHelper.IsMemoryAvailable())
-            {
-                CompilerServerLogger.Log("Memory available - accepting connection");
-                var clientLoggingIdentifier = $"Client{_clientLoggingIdentifier++}";
-                return new NamedPipeClientConnection(pipeStream, clientLoggingIdentifier);
-            }
+                    // The WaitForConnectionAsync API does not fully respect the provided CancellationToken
+                    // on all platforms:
+                    //
+                    //  https://github.com/dotnet/runtime/issues/40289
+                    //
+                    // To mitigate this we need to setup a cancellation Task and dispose the NamedPipeServerStream
+                    // if it ever completes. Once all of the NamedPipeServerStream for the given pipe name are
+                    // disposed they will all exit the WaitForConnectionAsync method
+                    var connectTask = pipeStream.WaitForConnectionAsync(cancellationToken);
+                    if (!PlatformInformation.IsWindows)
+                    {
+                        var cancelTask = Task.Delay(TimeSpan.FromMilliseconds(-1), cancellationToken);
+                        var completedTask = await Task.WhenAny(new[] { connectTask, cancelTask }).ConfigureAwait(false);
+                        if (completedTask == cancelTask)
+                        {
+                            throw new OperationCanceledException();
+                        }
+                    }
 
-            pipeStream.Close();
-            throw new Exception("Insufficient resources to process new connection.");
+                    await connectTask.ConfigureAwait(false);
```

Comment:
Should this be in an else?

Reply:
No because in both cases, windows and unix, you need to always await the `connectTask` so that exceptions from it are realized.